### PR TITLE
make more args kw only (except 'dim')

### PR DESCRIPTION
**Pull request by mathause**
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

This makes many arguments keyword-only, except for `dim` to avoid `da.weighted(...).mean("lat", "lon")` (i.e. `da.weighted(...).mean(dim="lat", skipna="lon")`) which silently does the wrong thing. I am sure I forgot some and for some I was unsure so I left them as is.

Question: do we want an deprecation cycle? Currently it just errors for `da.weighted(...).mean("dim", True)`. Might be nice to do it, however, @dcherian if I am not mistaken you did this without a deprecation in #5950, e.g. for `da.mean` etc.?

```python
import numpy as np
import xarray as xr
air = xr.tutorial.open_dataset("air_temperature")
wgt = np.cos(np.deg2rad(air.lat))
air.weighted(wgt).mean("lat", "lon")
```

### CODE REVIEW

**Comment by max-sixty** [comment]
This is great, thanks @mathause .

From what I've seen in the wild, I do think we're going to get some breaks on some of these without a deprecation — e.g. I've rarely seen an `n` arg used for `diff` . What are your thoughts on the link that @gcaria posted in https://github.com/pydata/xarray/issues/5531? That might do the deprecation for us.

**Comment by mathause** [comment]
Sounds good - I'll add a deprecation.

We might also want to do this for https://github.com/pydata/xarray/blob/main/xarray/core/_reductions.py

**Comment by mathause** [comment]
Sounds good - I'll add a deprecation.

We might also want to do this for https://github.com/pydata/xarray/blob/main/xarray/core/_reductions.py

**Comment by dcherian** [comment]
> We might also want to do this for [https://github.com/pydata/xarray/blob/main/xarray/core/_reductions.py](https://github.com/pydata/xarray/blob/main/xarray/core/_reductions.py?rgh-link-date=2022-03-23T17%3A29%3A01Z)

:+1: I lost track of it given that PR was open for so long. IIRC one issue is that the order of args was different forr Dataset.reduce vs DatasetGroupBy.reduce (cc @Illviljan)

**Comment by mathause** [comment]
From https://github.com/pydata/xarray/pull/5950#issuecomment-974719669

> Keyword-only and some had the 3rd position arg as `keep_attrs` and in other cases it was `axis`.

**Comment by Illviljan** [comment]
Yes, mypy found some issues with the ordering so moving to kwargs-only was my preferred solution. 
It indeed was a breaking change but I didn't think `keep_attrs`, `axis`, etc were that popular arguments either so I was not expecting too many noticing the change.

**Comment by max-sixty** [comment]
I would vote to resurrect this! Deprecations are nice but I would vote to merge this without them given that it seems that we won't merge it with them... (We can do it on a minor release and announce it)

What do folks think?

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -5995,7 +5999,7 @@ def _copy_attrs_from(self, other):
             if v in self.variables:
                 self.variables[v].attrs = other.variables[v].attrs
 
-    def diff(self, dim, n=1, label="upper"):
+    def diff(self, dim, *, n=1, label="upper"):
```
Comment (headtr1ck) [inline xarray/core/dataset.py:6002]:
```suggestion
    def diff(self, dim, n=1, *, label="upper"):
```

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -3210,7 +3213,9 @@ def _title_for_slice(self, truncate: int = 50) -> str:
 
         return title
 
-    def diff(self, dim: Hashable, n: int = 1, label: Hashable = "upper") -> DataArray:
+    def diff(
+        self, dim: Hashable, *, n: int = 1, label: Hashable = "upper"
```
Comment (headtr1ck) [inline xarray/core/dataarray.py:3217]:
```suggestion
        self, dim: Hashable, n: int = 1, *, label: Hashable = "upper"
```

**Review by headtr1ck — APPROVED** [review]
I'm ok with meeting this.

Maybe relax it a bit for the diff method, I remember using `n` as a positional argument in the past (coming from Matlab).

Should get an entry in breaking changes in the what's new though.

**Comment by mathause** [comment]
Happy to revive this PR. We do have a `@_deprecate_positional_args` helper - I added this as well (IIRC it was added after opening this PR). I am sure we could make _even more_ arguments positional but I should be doing my actual job ;-)

File: xarray/tests/test_dataset.py [hunk 3]
```diff
@@ -5051,9 +5051,9 @@ def test_dropna(self) -> None:
         ):
             ds.dropna("foo")
         with pytest.raises(ValueError, match=r"invalid how"):
-            ds.dropna("a", how="somehow")  # type: ignore
+            ds.dropna("a", how="somehow")
```
Comment (mathause) [inline xarray/tests/test_dataset.py:5054]: Not entirely sure if/ how these are related with the PR.

**Comment by mathause** [comment]
I think the failure is a fluke. Open/ close

**Review by dcherian — APPROVED** [review]
Thanks @mathause 

Let's add a whats-new note and merge.

File: xarray/tests/test_dataset.py (hunk 3 above)
Comment (max-sixty) [reply xarray/tests/test_dataset.py:5054]: Unfortunately my guess is that `@_deprecate_positional_args` doesn't pass types through, so these come as `Any`, so will need to be re-enabled when we remove the decorator...

Comment (mathause) [reply xarray/tests/test_dataset.py:5054]: I thought `@functools.wraps` conserves the `__signature__` of the wrapped function...

Comment (keewis) [reply xarray/tests/test_dataset.py:5054]: what it does (among other things) is set `wrapper.__wrapped__ = wrapped`. According to [this article]
(https://hynek.me/articles/decorators/) (which admittedly is a bit old), that doesn't mean the signature is the wrapped one's, just that `inspect.signature` is smart enough to display the wrapped signature.

In other words, it depends on how `mypy` fetches the type annotations.

Comment (headtr1ck) [reply xarray/tests/test_dataset.py:5054]: I think it should be easy to add typing to the deprecation helper. It should return `Callable[T, T]`, where `T = TypeVar("T", bound=Callable)`.

This means that mypy will complain if someone still uses positional arguments, which I think is exactly what we want.

Comment (mathause) [reply xarray/tests/test_dataset.py:5054]: I tried but feel free to push to the PR if it still need fixes.

File: xarray/util/deprecation_helpers.py [hunk 4]
```diff
@@ -71,7 +74,7 @@ def func(a, *, b=2):
     licences/SCIKIT_LEARN_LICENSE
     """
 
-    def _decorator(func):
+    def _decorator(func) -> Callable[[T], T]:
```
Comment (headtr1ck) [inline xarray/util/deprecation_helpers.py:77]:
```suggestion
    def _decorator(func: T) -> T:
```
This Typing should go to   `_deprecate_positional_args(version: str) -> Callable[[T], T]:` in line 47. Sorry for the confusion

**Review by headtr1ck — CHANGES_REQUESTED** [review]

File: xarray/tests/test_dataset.py (hunk 3 above)
Comment (headtr1ck) [reply xarray/tests/test_dataset.py:5054]: I'm mobile only currently, but I can see if I can fix it in the next two days.

Comment (mathause) [reply xarray/tests/test_dataset.py]: I think I got it now (maybe). But that means libraries that type check don't get a deprecation period (also fine by me, but then diverging from the runtime behaviour).

**Review by headtr1ck — APPROVED** [review]
Looks good now, thanks!

**Pull request merged**